# Two Hour Tour

So you're curious about Python but new to programming and unsure how it might help you in your research.  A great place to start!  This lesson is designed to be a 2 hour hands on tour of how Python can be used as a utility for processing text, particularly for humanities/literature projects.  Two hors isn't sufficient to give you every essential piece of knowledge, but it is enough time to better understand if it might be something useful to keep learning.

This means that we're going to dive right in and explore the basics as we're going.  Expect to not understand every piece of code you see here, and get comfortable with that.  Remember that we're going for a tour here rather than full tutorial.  You'd rather not invest days into learning Python only to realize that it isn't something useful for your projects, right?  But two hours is enough to explore some of the patterns and get some tools.

## Install stuff

This tutorial will somewhat presume that you can install Python on your own, but if you can't the repl provided below should be sufficient for this workshop.

### You can install things on your computer

You can follow the Anaconda and PyCharm instructions here:  https://github.com/elliewix/Ways-Of-Installing-Python/blob/master/ways-of-installing.md

1.  Install anaconda:  [link](https://github.com/elliewix/Ways-Of-Installing-Python/blob/master/ways-of-installing.md#step-1--install-python-itself)
2.  Install PyCharm:  [link](https://github.com/elliewix/Ways-Of-Installing-Python/blob/master/ways-of-installing.md#pycharm)
3.  Test your anaconda installation:  [link](https://github.com/elliewix/Ways-Of-Installing-Python/blob/master/ways-of-installing.md#appendix-1-testing-your-python-installation)
4.  Connect your PyCharm to anaconda:  [link](https://github.com/elliewix/Ways-Of-Installing-Python/blob/master/ways-of-installing.md#appendix-2--hooking-pycharm-to-your-anaconda-environment)

### You can't install things on your computer

There are a list of alternatives here: [link](https://github.com/elliewix/Ways-Of-Installing-Python/blob/master/ways-of-installing.md#i-dont-wantcant-install-things-on-this-computer)

I've also made a repl to go along with this course: https://repl.it/JE5R/3 This does presume that you have steaady internet access for the workshop, but easier than installing things.

# The core starting place: reading in data

We're going to jump right into a large block of code used to read in the contents of a text file.  Don't worry! We're going to discuss it in chunks, but there are still going to be chunks you won't (and don't need to) understand.

Here's the block of code to start with.  You should have this in your repl to run, or you can type it into your Python interpreter.

In [23]:
with open('pettigrew_letters_ORIGINAL.txt', 'r') as file_in:
    text = file_in.read()
    
print(len(text))

110828


**Key concept:  what is a script and how to run them**

This block of code is actually a complete Python script.  This means you can save this block of text and have python execute it from start to finish. You can do this inside of the repl you have access to.  That is a python script an when you click "run", it executes the script and prints out the results to the screen to the right.

Start first by looking at the output.

Question: What does the output "110828" represent?

The are a few key things that we can recognize in the code there:
* the file name
* and then there's a collection of clues like variable names that say "file_in" and methods & functions that say "read" and "len".  
* looking at the output number and scanning the contents of the file, we can see that it is too long to be the number of pages or lines in the text, and likely too big to be a represenation of the size (unless I've done something strange with the units).

Perhaps this is the number of characters in the document?

But at this point, we don't even know what this `text` thing is.  Let's just print out the text to start.|

**Key concept:  `.read()` will read in the entire contents of a file and save it to a variable**

In [8]:
print(text[:1000]) # for the purposes of this notebook, I've truncated the print to be the first 1000 characters

Pettigrew papers from the Beinecke Rare Book and Manuscript Library 
Series OSB MSS 113

Transcribed by Gabriel Moshenska (gmoshenska@yahoo.co.uk) and released under a CC-By license. This data is further described in the following data paper:
Moshenska, G. 2012. Selected Correspondence from the papers of Thomas Pettigrew (1791-1865), Surgeon and Antiquary. Journal of Open Archaeology Data 1(2), DOI: http://dx.doi.org/10.5334/4f913ca0cbb89

Letters are listed by location (box, folder) and are arranged alphabetically by writer.  All letters are addressed to Thomas Joseph Pettigrew (hereafter TJP).  

For further details see collection catalogue: http://drs.library.yale.edu:8083/fedora/get/beinecke:pettis1/PDF



List of letters:

Box 1, folder 10, William Francis Ainsworth to TJP, 17 December 1845
Box 1, folder 12, John Yonge Akerman to TJP, 13 November 1850
Box 1, folder 12, John Yonge Akerman to TJP, 7 March 1857
Box 1, folder 16, Thomas Amyot to TJP, 23 March 18??
Box 1, folder 16, Th

**Key concept:  you need to print stuff to your screen so your human body can view the contents**

We say that we "print" something to the screen for two reasons:
* that's what the function name is called, so that kind of makes sense
* computers didn't originally have electronic screens like we know now, so they literally printed stuff onto paper, and the name has stuck

You'll need to use the `print` function to print things to the screen to make them appear to you, the human.  Python sees the contents just fine, but you have to tell the compjuter what you want to appear on the screen.  NB: things get a bit different with this when you're in the interactive/interpreter envirionment.  

**key concept:  you "pass" functions values for them to take as input, and they do something with that or to that value.  What it does depends on the function**

So let's print out the contents of our `text` variable.  We do this by placing our variable name inside of the `()` for print.  That is actually called "passing".  We pass functions values inside their ().  You may also hear this referred to as parameters.  For our purposes, you can consider these synonyms.

At the bottom of your script, leaving everything else alone, add this line:  `print(text)`

**Key concept:  variables hold values for things and become placeholders for us to act on those values or data objects**

* Question:  This is a variable that holds a value. What is the current value of this variable?
* Answer:  The text contents of the 'pettigrew.txt' file.

We're going to expand on the idea of variables as we go.

## Let's back up for a moment and talk about the rest of the block of code you've just run

``` Python
with open('pettigrew_letters_ORIGINAL.txt', 'r') as file_in:
    text = file_in.read()
```

This block can be read as "read in the file `pettigrew_letters_ORIGINAL.txt` as text, and save the entire contents to the variable called `text`.  Pinning down exactly which parts of that sentence match up to the code is not exacltly a 1:1 activity, but we can look at some of the core elements of this pattern.

Python is a very high level language, meaning that it is further removed from 'computer speak' than many languages.  However, the act of reading in files is something with a lot of nuance and moving pieces that it usually exposes some of the lower level components of any language.


**Key concepts:  know the parts of speech in Python syntax, including keywords, variables, punctuation, indent, functions, and methods**

It is composed of:

* Keywords:  `with` and `as`
    * these help construct the `with` block of code (keep reading for more about blocks) by providing a template for us to plug in our desired stuff. 
    * imagine these as a madlib, "the \_\_\_ ate the \_\_\_" you fill in the subject and the object
    * so `with` and `as` provide a little mad lib for `with (*open function for the thing you want*) as (*file handling variable*).  You provide the contents of what you want to open, and provide the name of the file handling variable.  This line is then finished with a `:` to indicate you are done opening the `with` block.
* the literal string values for the file name and the open mode
* the indent:  see that white space on line two?
    * that extra white space is how you define with block of code that belongs to the `with` statement.
    * some languages use braces of various kinds to define how code relates to blocks, but in this case Python uses white space.
* variables:  `file_in` and `text`
    * these items both store values and represent a data type that can be acted on in certain ways
* functions and methods:  `open()` and `.readlines()`
    * these are the things that end with `()` and take specific actions.  You can call functions as part of a normal code, such as how you've seen `print`, `len`, and `open`.  
    
**key concept:  piecing together the meaning of a block of code based on your understanding of the moving parts**

So I read this block of code as:

* I'm opening the file (the `open()` function) in read mode (the `r` parameter tells me this)
* that open file object is being saved as `file_in`. This is the object that can be acted on the read the contents
    * there's a distinction between a file being "open for reading" versus "being read".
    * being "open for reading" means that Python knows what the file is, knows it is there, and has an internal connection to it
    * "being read" means I'm actively accessing and processing the contents of that file
    * this is normally a single step in programs, so a bit funny when out of GUI context
* I'm calling `.read()` on the file object, so I know that the full contents of that file will be read and stored into the variable `text`
* I can see that the `with` syntax is being used here, so I don't need to call `file_in.close()` to close the connection to the file.
    * I know this because I know the `with` method because I also learned about it in a workshop
    
Learning how to read code is a skill that you develop over time as you learn to recognize components and patterns.  Remember that you had to learn how to read a paragraph, and you did so by reading words into sentences, and sentences into narrative thought.  Code is the same way, but you don't always read it in such a linear process.

**key concept:  finding the data type of an object with `type()`**

Next, change `print(text)` to `print(type(text))`.

* Question: What does it tell you, and what do you think it means?
* Answer: it says `'str'` which stands for string.  So this just means that the `text` object holds some text in it.


**key concept:  you can nest functions and values**

Let's also take a moment to unpack the syntax we've seen of `print(len(text))` and `print(type(text))`.

These parentheses `()` have significant meaning, in that they hold a value to pass to the function.  This means that the function, in this case `print` takes an input of whatever you pass it.  In the two cases with `len` and `type` they are also functions, but they are being nested within another function.

Let's go back to the top of our code and consider the second line where it says `text = file_in.read()`.  Change that `read` to `readlines()`.  So your full script should now have the following:

In [10]:
with open('pettigrew_letters_ORIGINAL.txt', 'r') as file_in:
  text = file_in.readlines()
  
print(len(text))
print(type(text))

1911
<class 'list'>


Our two values have how changed.  What do you think the `1911` stands for?

How can you check?

Answers:

It represents the number of lines found within the tex file.


**key concept:  use `.readlines()` to read in a text file and store each line (as defined by a newline character (`\n`) as a string inside of a list**

Add `print(text)` to the bottom of your script and rerun it.

In [14]:
print(text[:50]) # for the purposes of keeping this notebook small, I've limited this to the first 50 lines

['Pettigrew papers from the Beinecke Rare Book and Manuscript Library \n', 'Series OSB MSS 113\n', '\n', 'Transcribed by Gabriel Moshenska (gmoshenska@yahoo.co.uk) and released under a CC-By license. This data is further described in the following data paper:\n', 'Moshenska, G. 2012. Selected Correspondence from the papers of Thomas Pettigrew (1791-1865), Surgeon and Antiquary. Journal of Open Archaeology Data 1(2), DOI: http://dx.doi.org/10.5334/4f913ca0cbb89\n', '\n', 'Letters are listed by location (box, folder) and are arranged alphabetically by writer.  All letters are addressed to Thomas Joseph Pettigrew (hereafter TJP).  \n', '\n', 'For further details see collection catalogue: http://drs.library.yale.edu:8083/fedora/get/beinecke:pettis1/PDF\n', '\n', '\n', '\n', 'List of letters:\n', '\n', 'Box 1, folder 10, William Francis Ainsworth to TJP, 17 December 1845\n', 'Box 1, folder 12, John Yonge Akerman to TJP, 13 November 1850\n', 'Box 1, folder 12, John Yonge Akerman to TJP, 7 Ma

What can we observe about this output?

* the text file content is there
* but it is broken up by line
* the lines all appear in the correct order that matches the original file
* none of the line breaks are printing
* there's the `\n` character showing up (hint: there's our newline, it's a text character!)
* each line is surrounded by `''`
* all the lines are contained within `[]` (this is our list, as you can see in the `print(type(text))` output

**key concept: lists are containers (bound by square brackets `[]`) that can store values and objects and have a specific order**

**key concept:  there are several ways to read in files, and you can change them by calling the different read methods on the file reading object.**

So let's hone in on the piece of code that we changed.

`file_in.read()` -> `file_in.readlines()`

**key concept:  the `.` notation allows us to call methods/functions on specific objects.  That object "knows" that it can do certain tasks, and you access those tasks via a `.` followed by the name of the method.**

Our `file_in` variable remained the same, but we changed the action that we took on it.  More speficially, we called `readlines()` on it rather than `read()`.  The `.` there matters a lot (R people, let's have this conversation).  


**key concept:  some functions will work on many kinds of objects, but they usually act differently when they do.**

We can also note that our functions of `len` and `type` still worked on `text` even though the content and data type of `text` has changed.  We can further observe that `len` will out the number of characters in a string while counting the number of elements in a list.

Let's change this one more time:  change `file_in.readlines()` to `file_in.readline()` (so you're taking the s off and making it singular).  Now run your script again.

In [16]:
with open('pettigrew_letters_ORIGINAL.txt', 'r') as file_in:
  text = file_in.readline()
  
print(len(text))
print(type(text))

69
<class 'str'>


Hmm, that's much shorter. What do you think happened?  Let's print out `text` to investigate.

In [17]:
print(text)

Pettigrew papers from the Beinecke Rare Book and Manuscript Library 



What can we observe about this output?

* Just one line of text
* the data type is back to a string
* look closely at the output, there's an extra line in there and the `\n` is gone. This indicates that the `\n` was rendered by the print statement instead of being printed out in the raw format.

**key concept:  the `\n` character is the newline character.  When rendered on the screen it will look like an empty line, but will appear as `\n` elsewhere.  It can also be directly typed in using `\n`.**

In [18]:
with open('pettigrew_letters_ORIGINAL.txt', 'r') as file_in:
    text = file_in.readline()
    text = file_in.readline()
  
print(len(text))
print(type(text))

19
<class 'str'>


In [20]:
print(text)

Series OSB MSS 113



Take a look at the actual text file.  What do you think has happened here?

Answer:  The first line was read in, as before, and assigned to `text`.  Then the second line was read in, and also assigned to the variable `text`.  This overwrote the original value.

So add the numbers 1 and 2 to the `text` variable name so you have `text1` and `text2`.  Now just have `print(text1)` and `print(text2)`.

**key concept:  `.readline()` will read one line at a time, which you can choose to only temporarily store.
Now add in another `file_in.readline()`.  Yup, copy/paste in another line of it and run the script again.**

In [22]:
with open('pettigrew_letters_ORIGINAL.txt', 'r') as file_in:
    text1 = file_in.readline()
    text2 = file_in.readline()
  
print(text1)
print(text2)

Pettigrew papers from the Beinecke Rare Book and Manuscript Library 

Series OSB MSS 113



## Module summary

* how to run a python script
* even though you may not understand everything in this block of code, you can know enough about what it is doing to make it useful.
* you can change around certain 'action' areas of the template code to customize what you want it to do
* you've explored several key methods of reading in text files.

# Let work with the file content now

Selecting the reading in method for your file is a personal/stylistic choice.  You can usually get away with using any of the three methods (`.read()`, `.readlines()`, or `.readline()`) but some will require more fussing than others depending on the task.  Your best bet is to try and match the task with the tool, but this is something you develop over time.

The Pettigrew file contains the plain text transcription of correspondance from Pettigrew's files.  These letters are identified with a box number and a folder number.  Together these should create a unique identifier for each letter. 

The file also has a manifest at the top that reports out which letters (again as specified by the corresponding box/folder numbers).

Here's the question:  are all the letters reported in the manifest actually there?

So here's two tasks:  

1. check if all the letters reported in the manifest appear in the actual file.
2. check if all the letters in the file are reported in the manifest

Discussion:  what are some approaches you can think of?  Take a moment outside of this notebook and look through the text file.

Key facts:

* Our most common unit of data is the line of text.  This means that each line reporting the box/folder numbers are contained on a single line of text.
    * This tells us that reading in the file via `.readlines()` would be good
* Each line that is reporting a letter location starts with the literal word "Box " then followed by a number
* No line reporting a letter location has any other information
* A mix of "folder" and "folders" is used
* All location lines consistantly use Box (e.g. Boxes does not appear)

Once we start with these clues, we can then seek out how to solve the problem in Python.

Our next task is to think about what our end points should be.  Let's say we want to construct two variables: a list of every location line from the manifest, and a list of every location line within the file proper.

So we begin to move toward our goal!

## Step 1: read in the file using `.readlines()`

We know this code well by now!

In [24]:
with open('pettigrew_letters_ORIGINAL.txt', 'r') as file_in:
    text = file_in.readlines()
    
print(len(text))

1911


We can look in our file and see that there are 1911 lines in the file, so looks good so far!

### Working with lists

We've got our data stored in a list, and we need to learn how to perform inquiries into a list.

Recall that lists have a specific order about them, so we can look up the values by those places.  This is called the index position.  Index positions start at 0.  No, I'm not going to defend this.  But it is reality that we have to work with.

The basic syntax for looking up an item in a list by the index position is:

`list_name[position_number]`

Take a moment to experiment with this in your script.

**Key concept: you can provide an index number to look up the value of something in a list based on that position**

Here's some starting points:

* `text[0]` will give you the first line
* `text[4]` will give you the fifth line (I know)
* `text[-1]` will give you the last line

**Key concept:  integer numbers are a data type, and they represent whole numbers either positive or negative**

Something to note about these numbers is that they are just the plain numbers.  No quotes or anything else.  These are what are called integer numbers, or whole numbers.

In [29]:
print("Line 1 is:", text[0])
print("Line 5 is:", text[4])
print("The last line is:", text[-1])

Line 1 is: Pettigrew papers from the Beinecke Rare Book and Manuscript Library 

Line 5 is: Moshenska, G. 2012. Selected Correspondence from the papers of Thomas Pettigrew (1791-1865), Surgeon and Antiquary. Journal of Open Archaeology Data 1(2), DOI: http://dx.doi.org/10.5334/4f913ca0cbb89

The last line is: 18th June 1833



Take a look at the contents of the `print` functions there.  I've passed in a string as well to improve readability.  This means that you can pass multiple things into the `print` function!  You actually can pass as many things as you want in there, just separate them with a comma.  You can even mix data types, such as strings and numbers.  Bonus, it'll automatically put that space in there for you.

**key concept:  you can pass multiple things into a single `print()` function to have them all print in one line.**

## Step 2: Use list slicing to get the manifest list out

Sometimes we also want to get groups of elements out of a list. We call this "slicing".  We can take a "slice" from a list using the following notation, which we'll unpack.

This is one of those syntax patterns worth just writing down somewhere prominent and referencing a lot.

`my_list[inclusive_start_position:exclusive_stop_position:optional_step_amount]`

A few things to note about the punctuation here:

* we add a set of square brackets `[]` after our list's variable name
* then we add our position numbers in separated by a colon character `:`

Now, let's unpack the names:

* inclusive_start_position:  this position number will be included, so if I say 4, then the item at position 4 will be included
* exclusive_stop_position:  this position number will NOT be included, so if I say 10, then it will go up to but not include item 10, so stopping at including 9.
* optional_step_amount:  this is an optional thing to include, and it's just the number of items to move forward, e.g. if I said 2, it would give me every other item in the list

**key concept:  you can slice out elements within a list using the start:stop:step notation**

Technically speaking, all these numbers are optional.  This statement will make more sense with examples.  Take some time to play with these in pairs. Fun fact: you can use the negative position numbers in here, what will that do?

In [30]:
print(text[0:3]) # will print a list with lines 0, 1, and 2 but NOT 3

['Pettigrew papers from the Beinecke Rare Book and Manuscript Library \n', 'Series OSB MSS 113\n', '\n']


In [31]:
print(text[5:6]) # will print a list with only line 5

['\n']


In [32]:
print(text[0:10:2]) # will print a list with lines 0, 2, 4, 6, and 8

['Pettigrew papers from the Beinecke Rare Book and Manuscript Library \n', '\n', 'Moshenska, G. 2012. Selected Correspondence from the papers of Thomas Pettigrew (1791-1865), Surgeon and Antiquary. Journal of Open Archaeology Data 1(2), DOI: http://dx.doi.org/10.5334/4f913ca0cbb89\n', 'Letters are listed by location (box, folder) and are arranged alphabetically by writer.  All letters are addressed to Thomas Joseph Pettigrew (hereafter TJP).  \n', 'For further details see collection catalogue: http://drs.library.yale.edu:8083/fedora/get/beinecke:pettis1/PDF\n']


Now that we can slice out stuff, how about we go after the letter manifest?  

There are programmatic ways to sort out where it begins and ends, but we can also just hard code it.

Challenge:  look through the file with a partner and determine the list slice that will give you the manifest.  Save that new list to a variable called `manifest_list` and print the length.  You should end up with a list of lenth 144.

In [36]:
manifest_list = text[15:159]
print(len(manifest_list))

144


## Step 3: loop over the list of manifest letters

Now that we have our data in a collection, we can loop through it.  This means that we tell Python to traverse that list of data, giving us access to each data point one at a time.

Accessing the data one at a time allows us to write a short piece of code that is generalized across the entire loop.  We will use an interable variable to reference the changing data items (remember, one at a time) inside this loop.  

Since this is a loop over a specific set of items, it is a definite loop.  Meaning that know (or can know) exactly how many times it should run.

There are two parts to a for loop:

1. The setup (remember how we had the setup line for the `with` statement?) where you define what you're looping over and the name of the iterable variable.
2. The block of code you'd like executed for each item in that collection or sequence.

**key concept:  for loops allow you to act on all the items in a collection, one at a time**

So I already have my list of letters in the manifest, called `manifest_list`.  I need to pick out the name of my iterable variable.  This will hold the location value of each letter in succession.  So let's call it `location`.  Thus, the opening line for my for loop would be:

`for location in manifest_list:`

See the madlib there?  `for (iterable_variable) in (list or other sequence):` . I need to provide the thing to iterate over, and then name what the iterable variable should be called.  

The iterable variable will be created during the first run of the loop.  Don't worry, it's completely fine to not have ever mentioned this variable anywhere else.  In fact, you should always try to come up with unique iterable variable names for each loop.  You can accidentially erase an existing variable's value if you reuse that name.  This variable will also persist after the for loop is done, holding the last value it saw before exiting from the for loop.  Occasionally this is con

**key concept:  iterable variables are created during the for loop, but will continue to exist after. Don't reuse variable names!**

In [39]:
for location in manifest_list:
    num_times = text.count(location)
    print(location, num_times)

Box 1, folder 12, John Yonge Akerman to TJP, 13 November 1850
 1
Box 1, folder 12, John Yonge Akerman to TJP, 7 March 1857
 1
Box 1, folder 16, Thomas Amyot to TJP, 23 March 18??
 1
Box 1, folder 16, Thomas Amyot to TJP, 20 August 18??
 1
Box 1, folder 16, Thomas Amyot to TJP, n.d.
 1
Box 1, folder 16, Thomas Amyot to TJP, 24 August 18??
 1
Box 1, folder 20, Francis Arundale to TJP, n.d.
 1
Box 1, folder 27, Robert John Eden Auckland to TJP, 20 June 18??
 1
Box 1, folder 28, William Ayrton to TJP, 1 February 1827
 1
Box 1, folder 29, Benjamin Guy Babbington to TJP, 10 June 18??
 1
Box 1, folder 31, W Baird to TJP, 22 May 18??
 1
Box 1, folder 33, Charles Baring, Bishop of Durham to TJP, n.d.
 1
Box 1, folder 37, Thomas Bateman to TJP, 4 July 1850
 1
Box 1, folder 37, Thomas Bateman to TJP, 18 January 1851
 1
Box 1, folder 37, Thomas Bateman to TJP, 29 June 1852
 1
Box 1, folder 39, William Beattie to TJP, 17 June 1829
 1
Box 1, folder 51, Sir William Betham to TJP, 1 June 1826
 1
Box 1